### PR TITLE
KAFKA-5150 reduce lz4 decompression overhead

### PR DESCRIPTION
**Pull request by xvrl**
- reuse decompression buffers in consumer Fetcher
- switch lz4 input stream to operate directly on ByteBuffers
- avoids performance impact of catching exceptions when reaching the end of legacy record batches
- more tests with both compressible / incompressible data, multiple
  blocks, and various other combinations to increase code coverage
- fixes bug that would cause exception instead of invalid block size
  for invalid incompressible blocks
- fixes bug if incompressible flag is set on end frame block size

Overall this improves LZ4 decompression performance by up to 40x for small batches.
Most improvements are seen for batches of size 1 with messages on the order of ~100B.
We see at least 2x improvements for for batch sizes of < 10 messages, containing messages < 10kB

This patch also yields 2-4x improvements on v1 small single message batches for other compression types.

Full benchmark results can be found here
https://gist.github.com/xvrl/05132e0643513df4adf842288be86efd

### CODE REVIEW

**Review by ijuma — COMMENTED** [review]
Thanks for the PR, great results! Why not include the benchmarks in the jmh-benchmarks module?

**Comment by xvrl** [comment]
@ijuma sure, I can add the benchmarks, wasn't aware of the module

File: clients/src/main/java/org/apache/kafka/common/record/AbstractLegacyRecordBatch.java [hunk 1]
```diff
@@ -291,12 +290,12 @@ public AbstractLegacyRecordBatch nextBatch() throws IOException {
...
     }
 
-    private static class DeepRecordsIterator extends AbstractIterator<Record> implements CloseableIterator<Record> {
+    protected static class DeepRecordsIterator extends AbstractIterator<Record> implements CloseableIterator<Record> {
         private final ArrayDeque<AbstractLegacyRecordBatch> batches;
         private final long absoluteBaseOffset;
         private final byte wrapperMagic;
 
-        private DeepRecordsIterator(AbstractLegacyRecordBatch wrapperEntry, boolean ensureMatchingMagic, int maxMessageSize) {
+        protected DeepRecordsIterator(AbstractLegacyRecordBatch wrapperEntry, boolean ensureMatchingMagic, int maxMessageSize) {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/record/AbstractLegacyRecordBatch.java]: Are the `private` -> `protected` changes still needed?

File: clients/src/main/java/org/apache/kafka/common/record/CompressionType.java [hunk 2]
```diff
@@ -96,9 +117,9 @@ public OutputStream wrapForOutput(ByteBufferOutputStream buffer, byte messageVer
         }
 
         @Override
-        public InputStream wrapForInput(ByteBufferInputStream buffer, byte messageVersion) {
+        public InputStream wrapForInput(ByteBuffer buffer, byte messageVersion) {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/record/CompressionType.java]: For consistency, it may make sense for `wrapForOutput` to take a `ByteBuffer` too.

File: clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java [hunk 3]
```diff
@@ -210,8 +209,7 @@ public int partitionLeaderEpoch() {
     private CloseableIterator<Record> compressedIterator() {
         ByteBuffer buffer = this.buffer.duplicate();
         buffer.position(RECORDS_OFFSET);
-        final DataInputStream stream = new DataInputStream(compressionType().wrapForInput(
-                new ByteBufferInputStream(buffer), magic()));
+        final DataInputStream stream = new DataInputStream(compressionType().wrapForInput(buffer, magic()));
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java]: We could change `DefaultRecord.readFrom` to take a `DataInput` (instead of `DataInputStream`), change `KafkaLZ4BlockInputStream` to implement `DataInput` and then only wrap if the returned value is not already a `DataInput`. That would remove a layer of indirection and if it's possible to implement `readFully` more efficiently in `KafkaLZ4BlockInputStream`, then it could be a win. If you have time, it may be worth a try.

File: clients/src/main/java/org/apache/kafka/common/record/KafkaLZ4BlockInputStream.java [hunk 4]
```diff
@@ -16,46 +16,56 @@
...
-    private final XXHash32 checksum;
-    private final byte[] buffer;
-    private final byte[] compressedBuffer;
+    private static final LZ4SafeDecompressor DECOMPRESSOR = LZ4Factory.fastestInstance().safeDecompressor();
+    private static final XXHash32 CHECKSUM = XXHashFactory.fastestInstance().hash32();
+
+    private final ByteBuffer in;
+
+    private ByteBuffer theBuffer;
+    private final ByteBuffer decompressionBuffer;
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/record/KafkaLZ4BlockInputStream.java]: It wasn't clear to me why we needed both `theBuffer` and `decompressionBuffer`. It seems that we rely on the fact that `theBuffer` is only populated after the first `readBlock` to implement `available` correctly. Is that the only difference? If so, would it not be simpler to have a single buffer that is allocated for the first time in `readBlock`?

File: clients/src/main/java/org/apache/kafka/common/record/KafkaLZ4BlockInputStream.java [hunk 5]
```diff
@@ -260,5 +292,4 @@ public synchronized void reset() throws IOException {
     public boolean markSupported() {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/record/KafkaLZ4BlockInputStream.java:287]: `mark()` and `reset()` are synchronized and probably should not be.

File: jmh-benchmarks/src/main/java/org/apache/kafka/common/record/DeepRecordsIteratorBenchmark.java [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+    public enum Bytes {
+        RANDOM, ONES
+    }
+
+    @Param(value = {"LZ4", "SNAPPY"})
+    private CompressionType type = CompressionType.NONE;
+
+    @Param(value = {"1", "2"})
+    byte messageVersion = MAGIC_VALUE_V1;
```
Comment (ijuma) [inline jmh-benchmarks/src/main/java/org/apache/kafka/common/record/DeepRecordsIteratorBenchmark.java]: I think the default should be `CURRENT_MAGIC_VALUE`.

File: jmh-benchmarks/src/main/java/org/apache/kafka/common/record/DeepRecordsIteratorBenchmark.java [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    final int batchCount = 5000;
+
+    private int maxBatchSize = 10;
+
+    public enum Bytes {
+        RANDOM, ONES
+    }
+
+    @Param(value = {"LZ4", "SNAPPY"})
+    private CompressionType type = CompressionType.NONE;
```
Comment (ijuma) [inline jmh-benchmarks/src/main/java/org/apache/kafka/common/record/DeepRecordsIteratorBenchmark.java]: Why did you remove `NONE` from this list? It seems useful to have the uncompressed baseline.

File: jmh-benchmarks/src/main/java/org/apache/kafka/common/record/DeepRecordsIteratorBenchmark.java [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+
+import static org.apache.kafka.common.record.RecordBatch.MAGIC_VALUE_V1;
+
+@State(Scope.Benchmark)
+public class DeepRecordsIteratorBenchmark {
+
+    final Random random = new Random(0);
+    final int batchCount = 5000;
+
+    private int maxBatchSize = 10;
```
Comment (ijuma) [inline jmh-benchmarks/src/main/java/org/apache/kafka/common/record/DeepRecordsIteratorBenchmark.java]: Seems like more fields could be final in this class.

File: jmh-benchmarks/src/main/java/org/apache/kafka/common/record/DeepRecordsIteratorBenchmark.java [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+        for (int i = 0; i < batchCount; ++i) {
+            int size = random.nextInt(maxBatchSize) + 1;
+            batch[i] = createBatch(size);
+            this.batchSize[i] = size;
+        }
+    }
+
+    private ByteBuffer createBatch(int batchSize) {
+        byte[] value = new byte[messageSize];
+        final ByteBuffer buf = ByteBuffer.allocate((messageSize + 8 + 4) * batchSize);
```
Comment (ijuma) [inline jmh-benchmarks/src/main/java/org/apache/kafka/common/record/DeepRecordsIteratorBenchmark.java]: You should probably use `AbstractRecords.sizeInBytesUpperBound`.

File: jmh-benchmarks/src/main/java/org/apache/kafka/common/record/DeepRecordsIteratorBenchmark.java [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+            builder.append(0, null, value);
+        }
+        return builder.build().buffer();
+    }
+
+    @Fork(value = 1)
+    @Warmup(iterations = 5)
+    @Benchmark
+    public void measureSingleMessage(Blackhole bh) throws IOException {
+        for (Record record : new ByteBufferLogInputStream(theBuffer.duplicate(), Integer.MAX_VALUE).nextBatch()) {
```
Comment (ijuma) [inline jmh-benchmarks/src/main/java/org/apache/kafka/common/record/DeepRecordsIteratorBenchmark.java]: To make the benchmark less dependent on implementation details, it would be better if it used `MemoryRecords.readableRecords(...).batches()`. Same for the other benchmark.

File: jmh-benchmarks/src/main/java/org/apache/kafka/common/record/DeepRecordsIteratorBenchmark.java [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.Random;
+
+import static org.apache.kafka.common.record.RecordBatch.MAGIC_VALUE_V1;
+
+@State(Scope.Benchmark)
+public class DeepRecordsIteratorBenchmark {
```
Comment (ijuma) [inline jmh-benchmarks/src/main/java/org/apache/kafka/common/record/DeepRecordsIteratorBenchmark.java]: Maybe this should be `RecordBatchIterationBenchmark`?

File: clients/src/main/java/org/apache/kafka/common/record/CompressionType.java [hunk 12]
```diff
@@ -156,7 +177,8 @@ else if (LZ4.name.equals(name))
     private static class LZ4Constructors {
         static final MethodHandle INPUT = findConstructor(
                 "org.apache.kafka.common.record.KafkaLZ4BlockInputStream",
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/record/CompressionType.java]: We followed the same approach for Snappy and LZ4 to deal with the possibility that the libraries are not in the classpath because it's not supported in a given platform. However, we are using Kafka classes for the block input stream and block output stream. As such, we can probably reference the constructor directly. We won't invoke the constructor unless LZ4 is configured and it's OK to fail in that case.

Does that make sense?

**Review by ijuma — COMMENTED** [review]
I did a first pass and left a few comments.

File: checkstyle/import-control.xml [hunk 13]
```diff
@@ -50,6 +50,9 @@
     <allow pkg="org.apache.kafka.common.internals" exact-match="true" />
     <allow pkg="org.apache.kafka.test" />
 
+    <allow pkg="org.openjdk.jmh.annotations" />
+    <allow pkg="org.openjdk.jmh.infra" />
```
Comment (ijuma) [inline checkstyle/import-control.xml]: Why do we need this?

File: clients/src/main/java/org/apache/kafka/common/record/AbstractLegacyRecordBatch.java (hunk 1 above)
Comment (xvrl) [reply clients/src/main/java/org/apache/kafka/common/record/AbstractLegacyRecordBatch.java]: we don't need them anymore, indeed

File: clients/src/main/java/org/apache/kafka/common/record/CompressionType.java (hunk 2 above)
Comment (xvrl) [reply clients/src/main/java/org/apache/kafka/common/record/CompressionType.java]: sure

File: checkstyle/import-control.xml (hunk 13 above)
Comment (xvrl) [reply checkstyle/import-control.xml]: the benchmark is in the `org.apache.kafka.common.record` package. Now that we don't rely on package protected classes anymore, we can move it to `org.apache.kafka.jmh` and remove this.

File: clients/src/main/java/org/apache/kafka/common/record/KafkaLZ4BlockInputStream.java (hunk 4 above)
Comment (xvrl) [reply clients/src/main/java/org/apache/kafka/common/record/KafkaLZ4BlockInputStream.java]: `theBuffer` is also pointed directly to the input buffer when a block is not compressed to avoid copying bytes.

File: clients/src/main/java/org/apache/kafka/common/record/KafkaLZ4BlockInputStream.java (hunk 5 above)
Comment (xvrl) [reply clients/src/main/java/org/apache/kafka/common/record/KafkaLZ4BlockInputStream.java:287]: hmm, not sure why they were in the first place

Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/record/KafkaLZ4BlockInputStream.java:287]: Copy and paste, I bet.

File: checkstyle/import-control.xml (hunk 13 above)
Comment (ijuma) [reply checkstyle/import-control.xml]: Sounds good.

File: clients/src/main/java/org/apache/kafka/common/record/KafkaLZ4BlockInputStream.java (hunk 4 above)
Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/record/KafkaLZ4BlockInputStream.java]: I had noticed that, but didn't think through how that could work if we had a single buffer variable. OK, seems simplest to have two buffers. Can we add a comment explaining this?

File: clients/src/main/java/org/apache/kafka/common/record/CompressionType.java (hunk 2 above)
Comment (xvrl) [reply clients/src/main/java/org/apache/kafka/common/record/CompressionType.java]: seems like this is not as straightforward as it seems. MemoryRecordsBuilder relies on having access to the underlying ByteBuffer held by the ByteBufferOutputStream, which would break if the buffer is expanded in ByteBufferOutputStream

File: clients/src/main/java/org/apache/kafka/common/record/CompressionType.java (hunk 12 above)
Comment (xvrl) [reply clients/src/main/java/org/apache/kafka/common/record/CompressionType.java]: makes sense, as far as I know lz4 falls back to a pure-java version anyway, so it should be safe regardless (unless there are other platform issues that are unrelated to native code)

File: clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java (hunk 3 above)
Comment (xvrl) [reply clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java]: I'd prefer to do that in a follow-up PR in trunk, since I'd also like to backport this to 0.10.2.x

File: clients/src/main/java/org/apache/kafka/common/record/KafkaLZ4BlockInputStream.java (hunk 4 above)
Comment (xvrl) [reply clients/src/main/java/org/apache/kafka/common/record/KafkaLZ4BlockInputStream.java]: will do

File: clients/src/main/java/org/apache/kafka/common/record/CompressionType.java (hunk 12 above)
Comment (xvrl) [reply clients/src/main/java/org/apache/kafka/common/record/CompressionType.java]: note however that we store the LZ4SafeDecompressor as a static field, so loading the Kafka class would also trigger loading the lz4 classes.

File: jmh-benchmarks/src/main/java/org/apache/kafka/common/record/DeepRecordsIteratorBenchmark.java (hunk 11 above)
Comment (xvrl) [reply jmh-benchmarks/src/main/java/org/apache/kafka/common/record/DeepRecordsIteratorBenchmark.java]: fixed

File: jmh-benchmarks/src/main/java/org/apache/kafka/common/record/DeepRecordsIteratorBenchmark.java (hunk 8 above)
Comment (xvrl) [reply jmh-benchmarks/src/main/java/org/apache/kafka/common/record/DeepRecordsIteratorBenchmark.java]: fixed

File: jmh-benchmarks/src/main/java/org/apache/kafka/common/record/DeepRecordsIteratorBenchmark.java (hunk 7 above)
Comment (xvrl) [reply jmh-benchmarks/src/main/java/org/apache/kafka/common/record/DeepRecordsIteratorBenchmark.java]: When using `MemoryRecords.readableRecords(...)` we don't technically test the Compression.NONE "compression" method, but test a different code path, so it's not really comparable if you care about an upper bound on how fast decompression could be.

Comment (xvrl) [reply jmh-benchmarks/src/main/java/org/apache/kafka/common/record/DeepRecordsIteratorBenchmark.java]: I added it back so we could at least see what it means in practice though

File: jmh-benchmarks/src/main/java/org/apache/kafka/common/record/DeepRecordsIteratorBenchmark.java (hunk 6 above)
Comment (xvrl) [reply jmh-benchmarks/src/main/java/org/apache/kafka/common/record/DeepRecordsIteratorBenchmark.java]: fixed

File: jmh-benchmarks/src/main/java/org/apache/kafka/common/record/DeepRecordsIteratorBenchmark.java (hunk 9 above)
Comment (xvrl) [reply jmh-benchmarks/src/main/java/org/apache/kafka/common/record/DeepRecordsIteratorBenchmark.java]: good to know

File: jmh-benchmarks/src/main/java/org/apache/kafka/common/record/DeepRecordsIteratorBenchmark.java (hunk 10 above)
Comment (xvrl) [reply jmh-benchmarks/src/main/java/org/apache/kafka/common/record/DeepRecordsIteratorBenchmark.java]: changed.

File: clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java (hunk 3 above)
Comment (xvrl) [reply clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java]: I was also thinking, if we avoided readFully altogether, we could avoid the EOFException penalty when reaching the end of the batch on legacy records and get a further 3x improvement for small legacy batches.

File: jmh-benchmarks/src/main/java/org/apache/kafka/common/record/DeepRecordsIteratorBenchmark.java (hunk 7 above)
Comment (ijuma) [reply jmh-benchmarks/src/main/java/org/apache/kafka/common/record/DeepRecordsIteratorBenchmark.java]: It's not the same code path, but that's OK. The comparison I'm looking for is a bit like the one in the LZ4 GitHub page where it compares memcpy with lz4 decompression (https://github.com/lz4/lz4).

File: clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java (hunk 3 above)
Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java]: Hmm, `DataInputStream.readFully` only throws an exception if we ask it to read past the end of the InputStream. So supposedly, if we fix the underlying InputStream, it's enough either way. The following PR does that:

https://github.com/apache/kafka/pull/2025/files#diff-eaa7e4414f285da2ff8e4508456078d2L192

**Comment by xvrl** [comment]
retest this please

**Comment by xvrl** [comment]
retest this please

**Comment by xvrl** [comment]
Retest this please

**Comment by xvrl** [comment]
retest this please

**Comment by xvrl** [comment]
retest this please

**Comment by xvrl** [comment]
@ijuma adressed all the comments I believe, now just waiting on getting the build to pass.
@hachikuji you might be interested in reviewing this part https://github.com/apache/kafka/pull/2967/commits/97e3ce81cea2736232a9527dc6a8cbd29e366ccd

**Comment by xvrl** [comment]
retest this please

**Comment by xvrl** [comment]
Retest this please

**Comment by xvrl** [comment]
retest this please

**Comment by xvrl** [comment]
retest this please

**Comment by xvrl** [comment]
Retest this please

**Comment by hachikuji** [comment]
@xvrl Probably not necessary to get a full green build if the failures are known issues. On the other hand, if you're just trying to shake out some more transient failures so that we can add JIRAs, have at it!

File: clients/src/main/java/org/apache/kafka/common/record/AbstractLegacyRecordBatch.java [hunk 14]
```diff
@@ -264,26 +263,48 @@ static void writeHeader(DataOutputStream out, long offset, int size) throws IOEx
     }
 
     private static final class DataLogInputStream implements LogInputStream<AbstractLegacyRecordBatch> {
-        private final DataInputStream stream;
+
+        private static final int OFFSET_OFFSET = 0;
+        private static final int SIZE_OFFSET = 8;
+        private static final int BUFFER_SIZE = 8 // offset
+                                               + 4; // size
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/record/AbstractLegacyRecordBatch.java]: nit: this looks a little weird. could we just do
```java
private static final int BUFFER_SIZE = 12; // offset + size
```
Also, the name is a bit vague. Maybe `RECORD_HEADER_SIZE`?

File: clients/src/main/java/org/apache/kafka/common/record/AbstractLegacyRecordBatch.java [hunk 15]
```diff
@@ -264,26 +263,48 @@ static void writeHeader(DataOutputStream out, long offset, int size) throws IOEx
...
             this.maxMessageSize = maxMessageSize;
+            this.headerBuffer = ByteBuffer.allocate(BUFFER_SIZE);
         }
 
         public AbstractLegacyRecordBatch nextBatch() throws IOException {
             try {
-                long offset = stream.readLong();
-                int size = stream.readInt();
+                int n = 0;
+                while (n < BUFFER_SIZE) {
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/record/AbstractLegacyRecordBatch.java]: This block seems to be the same code as below. Perhaps we could move it into a `readFully` function in `Utils` (we have a couple of these for working with `FileChannel` already)?

**Comment by xvrl** [comment]
@hachikuji I give up 😃  I think I've have each Java/Scala combination succeed at least  once. Also haven't seen any new failures in the last few runs.

**Comment by xvrl** [comment]
retest this please

File: clients/src/main/java/org/apache/kafka/common/record/CompressionType.java (hunk 2 above)
Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/record/CompressionType.java]: Good point, can we please add a comment explaining the reason for the inconsistency?

File: clients/src/main/java/org/apache/kafka/common/record/CompressionType.java (hunk 12 above)
Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/record/CompressionType.java]: Yeah, but `KafkaLZ4BlockInputStream` itself should not be initialised until it's actually used. I made this change, used CompressionType.SNAPPY.wrap* methods while LZ4 was not in the classpath and it worked fine. The same exercise did not work for Snappy because `SnappyInputStream` was not in the classpath. Anyway, we can keep it as it is if you prefer as the performance impact is low when compared to other things (8ns instead of 4ns per construction).

File: clients/src/main/java/org/apache/kafka/common/record/CompressionType.java (hunk 2 above)
Comment (xvrl) [reply clients/src/main/java/org/apache/kafka/common/record/CompressionType.java]: done

File: clients/src/main/java/org/apache/kafka/common/record/CompressionType.java (hunk 12 above)
Comment (xvrl) [reply clients/src/main/java/org/apache/kafka/common/record/CompressionType.java]: ok, I've change it to construct the lz4 streams directly, and updated the comments accordingly.

**Comment by xvrl** [comment]
@ijuma I have addressed all your comments

**Comment by xvrl** [comment]
Retest this please

**Comment by xvrl** [comment]
Retest this please

File: clients/src/main/java/org/apache/kafka/common/record/KafkaLZ4BlockInputStream.java [hunk 16]
```diff
@@ -65,31 +77,21 @@
      * @param ignoreFlagDescriptorChecksum for compatibility with old kafka clients, ignore incorrect HC byte
      * @throws IOException
      */
-    public KafkaLZ4BlockInputStream(InputStream in, boolean ignoreFlagDescriptorChecksum) throws IOException {
-        super(in);
-        decompressor = LZ4Factory.fastestInstance().safeDecompressor();
-        checksum = XXHashFactory.fastestInstance().hash32();
+    public KafkaLZ4BlockInputStream(ByteBuffer in, BufferSupplier bufferSupplier, boolean ignoreFlagDescriptorChecksum) throws IOException {
         this.ignoreFlagDescriptorChecksum = ignoreFlagDescriptorChecksum;
+        this.in = in.order(ByteOrder.LITTLE_ENDIAN);
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/record/KafkaLZ4BlockInputStream.java]: Is it safe to do this on the received buffer? `DeepRecordsIterator` doesn't seem to duplicate the buffer before calling `CompressionType.wrapForOutput`.

File: clients/src/main/java/org/apache/kafka/common/record/KafkaLZ4BlockInputStream.java [hunk 17]
```diff
@@ -145,46 +154,72 @@ private void readHeader() throws IOException {
      * @throws IOException
      */
     private void readBlock() throws IOException {
-        int blockSize = ByteUtils.readUnsignedIntLE(in);
+        if (in.remaining() < 4) {
+            throw new IOException(PREMATURE_EOS);
+        }
+
+        int blockSize = in.getInt();
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/record/KafkaLZ4BlockInputStream.java:155]: I'm not totally sure why we were previously using a `readUnsigned` method and now we're not. For many of the other cases where we do this, it's easy to verify that the values we care about work either way. The ones that weren't totally clear to me were this and the block checksum cases. Can you please elaborate why the change is safe in those two cases?

File: clients/src/main/java/org/apache/kafka/common/record/KafkaLZ4BlockInputStream.java [hunk 18]
```diff
@@ -145,46 +154,72 @@ private void readHeader() throws IOException {
...
+        boolean compressed = (blockSize & LZ4_FRAME_INCOMPRESSIBLE_MASK) == 0;
+        blockSize &= ~LZ4_FRAME_INCOMPRESSIBLE_MASK;
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/record/KafkaLZ4BlockInputStream.java:157]: Hmm, so this bug meant that the `blockSize > maxBlockSize` check was not working, right? Also, did we just get lucky that the `blockSize == 0` check worked?

File: clients/src/test/java/org/apache/kafka/common/record/KafkaLZ4Test.java [hunk 19]
```diff
@@ -42,37 +50,135 @@
...
+        if (!close || (useBrokenFlagDescriptorChecksum && !ignoreFlagDescriptorChecksum)) return;
+
+        thrown.expect(IOException.class);
+        thrown.expectMessage(CoreMatchers.containsString("exceeded max"));
+
+        byte[] compressed = compressedBytes();
+        final ByteBuffer buffer = ByteBuffer.wrap(compressed).order(ByteOrder.LITTLE_ENDIAN);
+
+        int blockSize = buffer.getInt(7);
+        blockSize = (blockSize & LZ4_FRAME_INCOMPRESSIBLE_MASK) | (1 << 24 & ~LZ4_FRAME_INCOMPRESSIBLE_MASK);
```
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/common/record/KafkaLZ4Test.java:161]: Nit: is there a reason why we are modifying the existing block size instead of just setting the bad block size directly?

**Review by ijuma — COMMENTED** [review]
Sorry for the delay, I finally had a chance to complete my review. I left 4 comments questions in this PR and submitted a separate PR with an alternative to the thread local buffers:

https://github.com/apache/kafka/pull/3164

I would normally do the PR with your branch as the target, but because I rebased my branch on trunk (and fixed the merge conflicts), this would not have the desired effect. Please take a look at the other PR and we can decide what we should pull into this PR.

File: clients/src/main/java/org/apache/kafka/common/record/KafkaLZ4BlockInputStream.java (hunk 17 above)
Comment (xvrl) [reply clients/src/main/java/org/apache/kafka/common/record/KafkaLZ4BlockInputStream.java:155]: from reading the code, `ByteUtils.readUnsignedIntLE` is a misnomer and should probably be just be called `readIntLE`. Unlike `readUnsignedInt`, which returns a long and applies the proper bit mask, `readUnsignedIntLE` returns an integer and doesn't do anything other than shifting the bytes.

File: clients/src/main/java/org/apache/kafka/common/record/KafkaLZ4BlockInputStream.java (hunk 16 above)
Comment (xvrl) [reply clients/src/main/java/org/apache/kafka/common/record/KafkaLZ4BlockInputStream.java]: `DeepRecordsIterator` technically duplicates the buffer when calling [`wrapperRecord.value()`](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/record/AbstractLegacyRecordBatch.java#L310) and never reuses it, but I agree that could easily be missed or introduce odd bugs if that code were to be rewritten. I can add a `duplicate()` here just for safety.

Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/record/KafkaLZ4BlockInputStream.java]: Sounds good.

File: clients/src/main/java/org/apache/kafka/common/record/KafkaLZ4BlockInputStream.java (hunk 18 above)
Comment (xvrl) [reply clients/src/main/java/org/apache/kafka/common/record/KafkaLZ4BlockInputStream.java:157]: Indeed, the size check wasn't working, but at the end of the day, we would just thrown a different exception in case the block size exceeded the max.

Regarding the zero size block, we were just lucky that we never set the `LZ4_FRAME_INCOMPRESSIBLE_MASK` flag in `KafkaLZ4BlockOutputStream.writeEndMark`. I checked the spec and it didn't seem to specify whether the flag should be set or not in that case, so better be safe.

File: clients/src/test/java/org/apache/kafka/common/record/KafkaLZ4Test.java (hunk 19 above)
Comment (xvrl) [reply clients/src/test/java/org/apache/kafka/common/record/KafkaLZ4Test.java:161]: yes, to make sure it works in both cases when the incompressible flag is set and when it is not.

File: clients/src/main/java/org/apache/kafka/common/record/CompressionType.java [hunk 20]
```diff
@@ -179,4 +185,17 @@ private static MethodHandle findConstructor(String className, MethodType methodT
...
+
+        ByteBuffer theBuffer;
+
+        @Override
+        public ByteBuffer get(int size) {
+            if (theBuffer == null || theBuffer.capacity() < size) {
+                theBuffer = ByteBuffer.allocate(size);
+            }
+            theBuffer.limit(size);
+            return theBuffer;
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/record/CompressionType.java]: Should we rewind the buffer in case the position wasn't reset after the last use?

(Also nit: can we just call it `buffer`?)

File: clients/src/main/java/org/apache/kafka/common/record/CompressionType.java [hunk 21]
```diff
@@ -179,4 +185,17 @@ private static MethodHandle findConstructor(String className, MethodType methodT
         }
     }
 
+    static class RecyclingBufferSupplier implements KafkaLZ4BlockInputStream.BufferSupplier {
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/record/CompressionType.java]: It's a little odd for this to live here given the interface is in `KafkaLZ4BlockInputStream`. I'd suggest either moving it there or pulling the `BufferSupplier` interface out of `KafkaLZ4BlockInputStream`.

**Review by hachikuji — COMMENTED** [review]
Thanks for the patch. I didn't look deeply at the changes in `KafkaLZ4BlockInputStream`, but I left a few comments in `AbstractLegacyRecordBatch`.

File: clients/src/main/java/org/apache/kafka/common/record/AbstractLegacyRecordBatch.java (hunk 14 above)
Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/record/AbstractLegacyRecordBatch.java]: Yeah, my PR removes all of these in favour of existing constants: https://github.com/apache/kafka/pull/3164/commits/a7bc3e6e98ad3f983f5619601b216e83142afe42

File: clients/src/main/java/org/apache/kafka/common/record/AbstractLegacyRecordBatch.java (hunk 15 above)
Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/record/AbstractLegacyRecordBatch.java]: You and I had the exact same thought: https://github.com/apache/kafka/pull/3164/commits/a7bc3e6e98ad3f983f5619601b216e83142afe42

File: clients/src/main/java/org/apache/kafka/common/record/CompressionType.java (hunk 21 above)
Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/record/CompressionType.java]: 3 for 3: https://github.com/apache/kafka/pull/3164/commits/8403cbb616dd88cd7b06dfe8910aa0d0cded8b6f#diff-ee888ddca9dfb15863eeddcd379f65c6R1

File: clients/src/main/java/org/apache/kafka/common/record/CompressionType.java (hunk 20 above)
Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/record/CompressionType.java]: That's one issue. The other is that if there are two `KafkaLZ4BlockInputStream` instances in the same thread: https://github.com/apache/kafka/pull/3164/commits/8403cbb616dd88cd7b06dfe8910aa0d0cded8b6f#diff-ee888ddca9dfb15863eeddcd379f65c6R36

Comment (xvrl) [reply clients/src/main/java/org/apache/kafka/common/record/CompressionType.java]: @hachikuji technically we don't have to, KafkaLZ4BlockInputStream resets position and limit every time anyway when preparing the buffer for consumption.

@ijuma agree that would be a concern if someone were to consume from two separate consumers in the same thread. Happy to replace the buffer supplier implementation with the one you pointed to.

Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/common/record/CompressionType.java]: @xvrl Yeah, figured that was the case, but it makes the code a bit brittle to depend on that.

**Comment by xvrl** [comment]
retest this please

**Comment by ijuma** [comment]
Client system tests build will appear in the following link when it starts: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/903/

**Comment by ijuma** [comment]
Clients system tests passed: https://jenkins.confluent.io/job/system-test-kafka-branch-builder-2/304/

**Review by ijuma — APPROVED** [review]
Thanks for the PR, this is a huge performance improvement. :)

**Comment by ijuma** [comment]
LGTM, merging to trunk and 0.11.0.
